File: Java_Joc_Windows_Android/AndroidStudioProject/core/src/base/view/listmenucardpopup/ListMenuCardPopupView.java
package base.view.listmenucardpopup;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;

import java.util.Map;

import base.interaction.TouchListener;
import base.menu.util.EaseConfig;
import base.menu.util.MenuItemConfig;
import base.menu.util.MenuItemConfig_ComposedVisual;
import base.menu.util.MenuItemConfig_Data;
import base.menu.util.MenuItemConfig_Visual;
import base.menu.util.MenuVisualAnimationProperty;
import base.menu.util.StateChangeDurations;
import local.app.AppLocal;
import base.util.AppStyles;
import base.visual.IDrawableNode;
import base.visual.IInteractiveDrawable;
import base.visual.util.VisualCreateConfig;
import base.visual.util.VisualPropertiesConfig;

public class ListMenuCardPopupView extends ListMenuCardPopupViewList {

    private TouchListener touchListener = new TouchListener() {
        @Override
        public boolean touchDown(IInteractiveDrawable target, float worldX, float worldY) {
            // Touch events are ignored while the menu is animating.
            if(!isActive()) {
                return false;
            }

            if(target == getItem(new String[] {"_root", "touchAreaFullScreen"})) {
                helperCallCloseRequestedEvent();
            }else if(target == getItem(new String[] {"_root", "contentArea", "touchAreaContentBg"})) {
                //void
            }else {
                Gdx.app.debug("Error", "Touch event received for a unknown target.");

                return false;
            }

            return false;
        }
    };

    private static final MenuItemConfig[] menuItemConfigs = new MenuItemConfig[] {
            new MenuItemConfig_Data("rendering3DSize", new VisualPropertiesConfig()
                    .put("width", 1f, new String[] {"_appSize"}, "width")
                    .put("height", 1f, new String[] {"_appSize"}, "height")
            ),

            new MenuItemConfig_Visual("layerBase", new String[] {"_root", "visualsHolder"}, VisualCreateConfig.newGroup(), new VisualPropertiesConfig()
                    .putAnimating("alpha", new MenuVisualAnimationProperty()
                            .putState("in", 1f, null, null, EaseConfig.DECELERATION)
                            .putState("out", 0f, null, null, EaseConfig.ACCELERATION)
                    )
            ),

            new MenuItemConfig_Visual("touchAreaFullScreen", new String[] {"_local", "layerBase"}, VisualCreateConfig.newTouchAreaAudible(null, null), new VisualPropertiesConfig()
                    .put("width", 1f, new String[] {"_local", "rendering3DSize"}, "width")
                    .put("height", 1f, new String[] {"_local", "rendering3DSize"}, "height")
                    .put("anchorX", 0.0f)
                    .put("anchorY", 0.0f)
                    .put("x", 0f)
                    .put("y", 0f)
                    //.put("isTouchFallthroughEnabled", true)
            ),

            new MenuItemConfig_Visual("screenDimmerFullScreen", new String[] {"_local", "layerBase"}, VisualCreateConfig.newSprite("visual/taUtilLinear.atlas", "bg/simple_white_FFFFFF"), new VisualPropertiesConfig()
                    .put("width", 1f, new String[] {"_local", "rendering3DSize"}, "width")
                    .put("height", 1f, new String[] {"_local", "rendering3DSize"}, "height")
                    .put("anchorX", 0.0f)
                    .put("anchorY", 0.0f)
                    .put("x", 0f)
                    .put("y", 0f)
                    .put("tintColor", Color.valueOf("#000000FF"))
                    .putAnimating("alpha", new MenuVisualAnimationProperty()
                            .putState(new String[] {"_in_", "_isFullScreenDimmer_"}, AppStyles.SCREEN_DIMMER_ALPHA, null, null, EaseConfig.DEFAULT)
                            .putState("out", 0f, null, null, EaseConfig.DEFAULT)
                    )
                    .putAnimating("_visibleFloat", new MenuVisualAnimationProperty()
                            .putState(new String[] {"_in_", "_isFullScreenDimmer_"}, 1f, null, null, EaseConfig.DEFAULT)
                            .putState("out", 0f, null, null, EaseConfig.DEFAULT)
                    )
            ),

            new MenuItemConfig_ComposedVisual("contentArea", new MenuItemConfig[] {
                    new MenuItemConfig_Visual("layerBase", new String[] {"_local", "parent", "layerBase"}, VisualCreateConfig.newGroup(), new VisualPropertiesConfig()
                            // Dynamic position
                    ),

                    menuItemConfigs_contentArea_optionsList,

                    new MenuItemConfig_ComposedVisual("shadowHolder", new MenuItemConfig[] {
                            new MenuItemConfig_Data("rendering3DSize", new VisualPropertiesConfig()
                                    .put("width", 1f, new String[] {"_local", "parent", "optionsList"}, "listWidth")
                                    .put("height", 1f, new String[] {"_local", "parent", "optionsList"}, "listHeight")
                            ),
                            new MenuItemConfig_Data("shadowOffset", new VisualPropertiesConfig()
                                    .put("x", 0f)
                                    .put("y", 0f)
                            ),
                            new MenuItemConfig_Visual("layerBase", new String[] {"_local", "parent", "layerBase"}, VisualCreateConfig.newGroup(), new VisualPropertiesConfig()
                                    .put("x", 0.5f, new String[] {"_local", "rendering3DSize"}, "width")
                                    .put("y", -0.5f, new String[] {"_local", "rendering3DSize"}, "height")
                            ),
                            AppStyles.ITEM_CONFIG_shadow_CARD
                    })
            })
    };

    private static final StateChangeDurations[] menuAnimationDurations = new StateChangeDurations[] {
            new StateChangeDurations("_out_", "_in_", AppStyles.ANIMATION_DURATION_S_DEFAULT),
            new StateChangeDurations("_in_", "_out_", AppStyles.ANIMATION_DURATION_S_DEFAULT)
    };

    public ListMenuCardPopupView(AppLocal app) {
        super(app, menuItemConfigs, menuAnimationDurations);

        //void
    }

    private void helperCallCloseRequestedEvent() {
        if(viewEventListener != null) {
            viewEventListener.closeRequested(this);
        }
    }

    @Override
    protected void createVisuals(Map<String, Object> createConfigMap) {
        super.createVisuals(createConfigMap);

        ((IInteractiveDrawable) getItem(new String[]{"_root", "touchAreaFullScreen"})).setTouchListener(touchListener);
    }

    @Override
    protected void activeStateChangedEvent() {
        super.activeStateChangedEvent();

        if(!isActive()) {
            //void
        }else {
            //void
        }
    }

    @Override
    protected void syncVisuals(boolean syncAbsoluteValues, boolean syncAnimatingValues, String startStateID, String endStateID, float stateCompletionPercent) {
        super.syncVisuals(syncAbsoluteValues, syncAnimatingValues, startStateID, endStateID, stateCompletionPercent);

        if(viewConfig != null) {
            IDrawableNode contentArea_layerBase = (IDrawableNode) getItem(new String[]{"_root", "contentArea", "layerBase"});

            contentArea_layerBase.setPosition(viewConfig.posXDip * app.screenDensityMultiplier.x, viewConfig.posYDip * app.screenDensityMultiplier.y);
        }
    }

}

File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/view/listmenucardpopup/ListMenuCardPopupView.cpp
#include "ListMenuCardPopupView.h"
#include <base/statics/StaticsInit.h>
#include <base/util/AppStyles.h>
#include <base/menu/util/MenuItemConfig_Data.h>
#include <base/menu/util/MenuItemConfig_Visual.h>
#include <base/menu/util/MenuItemConfig_ComposedVisual.h>
#include <base/visual2D/util/VCC_Group.h>
#include <base/visual2D/util/VCC_Sprite.h>
#include <base/visual2D/util/VCC_TouchArea.h>
#include <base/util/AppStyles.h>
#include <base/log/GlobalAccessLogHandler.h>
#include <base/app/IAppMetrics.h>

using namespace base;

ArrayList<MenuItemConfig*>* ListMenuCardPopupView::viewItemConfigs = base::StaticsInit::AddCbGeneral<ArrayList<MenuItemConfig*>*>(1, []() {
	viewItemConfigs = (new ArrayList<MenuItemConfig*>())
		->appendDirect_chain(new MenuItemConfig_Data("props", (new VisualPropertiesConfig())
				->put("width", 1.0f, "_appSize.width")
				->put("height", 1.0f, "_appSize.height")
		))

		->appendDirect_chain(new MenuItemConfig_Visual("layer", "_root.visualsHolder", VCC_Group::New(), (new VisualPropertiesConfig())
				->putAnimating("alpha", (new MenuVisualAnimationProperty())
						->putState("in", 1.0f, EaseConfig::DECELERATION)
						->putState("out", 0.0f, EaseConfig::ACCELERATION)
				)
		))

		->appendDirect_chain(new MenuItemConfig_Visual("touchAreaFullScreen", "layer", VCC_TouchArea::New(), (new VisualPropertiesConfig())
				->put("width", 1.0f, "_local.props.width")
				->put("height", 1.0f, "_local.props.height")
				->put("anchorX", 0.0f)
				->put("anchorY", 0.0f)
				->put("x", 0.0f)
				->put("y", 0.0f)
			//->put("isTouchFallthroughEnabled", true)
		))

		->appendDirect_chain(new MenuItemConfig_Visual("screenDimmerFullScreen", "layer", VCC_Sprite::New("visual2D/taUtilLinear.atlas", "bg/simple_white_FFFFFF"), (new VisualPropertiesConfig())
				->put("width", 1.0f, "_local.props.width")
				->put("height", 1.0f, "_local.props.height")
				->put("anchorX", 0.0f)
				->put("anchorY", 0.0f)
				->put("x", 0.0f)
				->put("y", 0.0f)
				->put("tintColor", Color("#000000FF"))
				->putAnimating("alpha", (new MenuVisualAnimationProperty())
						->putState("_in_._isFullScreenDimmer_", AppStyles::SCREEN_DIMMER_ALPHA, EaseConfig::LINEAR)
						->putState("out", 0.0f, EaseConfig::LINEAR)
				)
				->putAnimating("_visibleFloat", (new MenuVisualAnimationProperty())
						->putState("_in_._isFullScreenDimmer_", 1.0f, EaseConfig::LINEAR)
						->putState("out", 0.0f, EaseConfig::LINEAR)
				)
		))

		->appendDirect_chain(new MenuItemConfig_ComposedVisual("contentArea", (new ArrayList<MenuItemConfig*>())
			->appendDirect_chain(new MenuItemConfig_Visual("layer", "_local.parent.layer", VCC_Group::New(), new VisualPropertiesConfig())
				// Dynamic position
			)

			->appendDirect_chain(new_menuItemConfigs_contentArea_optionsList())

			->appendDirect_chain(new MenuItemConfig_ComposedVisual("shadowHolder", (new ArrayList<MenuItemConfig*>())
				->appendDirect_chain(new MenuItemConfig_Data("props", (new VisualPropertiesConfig())
					->put("width", 1.0f, "_local.parent.optionsList.listWidth")
					->put("height", 1.0f, "_local.parent.optionsList.listHeight")
				))
				->appendDirect_chain(new MenuItemConfig_Data("shadowOffset", (new VisualPropertiesConfig())
					->put("x", 0.0f)
					->put("y", 0.0f)
				))
				->appendDirect_chain(new MenuItemConfig_Visual("layer", "_local.parent.layer", VCC_Group::New(), (new VisualPropertiesConfig())
					->put("x", 0.5f, "_local.props.width")
					->put("y", -0.5f, "_local.props.height")
				))
				->appendDirect_chain(AppStyles::new_ITEM_CONFIG_shadow_CARD())
			))
		));
});

ArrayList<StateChangeDurations*>* ListMenuCardPopupView::viewAnimationDurations = base::StaticsInit::AddCbGeneral<ArrayList<StateChangeDurations*>*>(1, []() {
	viewAnimationDurations = (new ArrayList<StateChangeDurations*>())
		->appendDirect_chain(new StateChangeDurations("_out_", "_in_", AppStyles::ANIMATION_DURATION_S_DEFAULT))
		->appendDirect_chain(new StateChangeDurations("_in_", "_out_", AppStyles::ANIMATION_DURATION_S_DEFAULT));
	});

ListMenuCardPopupView::ListMenuCardPopupView(IApp* app, IHandlerAutoDisposal* handlerAutoDisposal)
	: super(app, viewItemConfigs, viewAnimationDurations, handlerAutoDisposal)
{
	touchListener.cb_onCursorDown = std::bind(&ListMenuCardPopupView::v2d_onCursorDown, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4, std::placeholders::_5);
}

bool ListMenuCardPopupView::v2d_onCursorDown(
	IInteractiveDrawable2D* target,
	int cursorIndex, base::Touch::ButtonCode* buttonId,
	Vector2& screenPos, Vector2& localPos)
{
	// Touch events are ignored while the menu is animating.
	if (!isActive()) {
		return false;
	}

	if (dynamic_cast<void*>(target) == dynamic_cast<void*>(touchAreaFullScreen)) {
		helperCallCloseRequestedEvent();
	} else {
		GlobalAccessLogHandler::devLogHandler->post_debug("Error", "Touch event received for a unknown target.");

		return false;
	}

	return false;
}

void ListMenuCardPopupView::helperCallCloseRequestedEvent() {
	if (listener != nullptr) {
		listener->onCloseRequested(this);
	}
}

void ListMenuCardPopupView::createVisuals(Map1D<std::string, void*>* createConfigMap) {
	super::createVisuals(createConfigMap);

	(touchAreaFullScreen = getItemMustExistAs<IInteractiveDrawable2D*>("_root.touchAreaFullScreen"))->addTouchListener(&touchListener);
}

void ListMenuCardPopupView::onSelectedStateChanged() {
	super::onSelectedStateChanged();

	if (!isActive()) {
		//void
	} else {
		//void
	}
}

void ListMenuCardPopupView::syncVisuals(bool syncAbsoluteValues, bool syncAnimatingValues, std::string startStateID, std::string endStateID, float stateCompletionPercent) {
	super::syncVisuals(syncAbsoluteValues, syncAnimatingValues, startStateID, endStateID, stateCompletionPercent);

	if (viewConfig != nullptr) {
		IDrawableNode2D* contentArea_layer = getItemMustExistAs< IDrawableNode2D * >("_root.contentArea.layer" );

		contentArea_layer->setPosition(
			viewConfig->posXDip * appMetrics->getScreen()->getScreenDensityMultiplier()->x, 
			viewConfig->posYDip * appMetrics->getScreen()->getScreenDensityMultiplier()->y
		);
	}
}

ListMenuCardPopupView::~ListMenuCardPopupView() {
	//void
}
